### PR TITLE
Add scroll buttons and auto-scroll to logs

### PR DESCRIPTION
**Pull request by MA9521**
Related to https://github.com/tektoncd/dashboard/issues/327

# Changes

First commit (auto-scroll):
Logs of step runs can get very big. While the step run has not terminated, the p
age will keep scrolling down the page as needed. This behaviour can be stopped once the user scrolls up so that the bottom of the log is no longer visible. This behaviour can be resumed once the user scrolls back down to the log bottom.

This feature is activated by a boolean flag (which is turned to true in the dashboard ui) not to affect the users of the reusable component `<Log` in edge cases (the component followed by other elements in the page,component inside a non-window scrollable parent.....)

Second commit( Scroll buttons):
Step-runs with huge logs require the user to manually scroll all the way down the page. A click/press-to-scroll-to-page-top button will now appear when the top of the logs is above the viewport and a click/press-to-scroll-to-page-bottom button will now appear when the bottom of the logs is below the viewport.

The button presence is updated at each Log component update or scroll. For a logically-complete solution, we would need to use a ResizeObserver. This feature is behind a flag to not affect the users of the reusable `<Log` component with edge cases (a non-window-scrollable parent).

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/main/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/main/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/main/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/main/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by AlanGreene** [comment]
@MA9521 and I spoke about this earlier today. We're going to try a different approach with `position: sticky` and `IntersectionObserver` to see if we can simplify the code while also accounting for the case where the scroll parent is not the window.

**Comment by AlanGreene** [comment]
Thanks for the PR @MA9521, I'll take a look at this today. Let me know if you have any problems with the CLA.

**Comment by MA9521** [comment]
Closing then will reopen to recheck CLA

File: packages/components/src/components/Log/Log.scss [hunk 1]
```diff
@@ -53,6 +53,20 @@ pre.tkn--log {
     right: 0;
   }
 
+  .bx--copy-btn.scroll-btn {
```
Comment (AlanGreene) [inline packages/components/src/components/Log/Log.scss:56]: I'm not seeing this style applied, instead the buttons just appear at the bottom of the log.

File: packages/components/src/components/Log/Log.js [hunk 2]
```diff
@@ -28,17 +30,147 @@ const itemSize = 15; // This should be kept in sync with the line-height in SCSS
...
+  constructor(props) {
+    super(props);
+    this.state = { loading: true };
+    this.logRef = React.createRef();
+    this.textRef = React.createRef();
+  }
 
   componentDidMount() {
     this.loadLog();
+    window.addEventListener("scroll", this.handleLogScroll, true);
```
Comment (AlanGreene) [inline packages/components/src/components/Log/Log.js:42]: Ideally we would only register the scroll listener when one of the features relying on it is enabled, similar with the other code here.

File: packages/components/src/components/Log/Log.js [hunk 3]
```diff
@@ -28,17 +30,147 @@ const itemSize = 15; // This should be kept in sync with the line-height in SCSS
...
+      || utils.hasElementPositiveScrollTop(this.textRef?.current?.firstElementChild);
+  };
+
+  scrollToBottomLog = () => {
+    const longTextElement = this.textRef?.current?.firstElementChild;
+    if (utils.hasElementPositiveScrollBottom(longTextElement)) {
+      longTextElement.scrollTop = longTextElement.scrollHeight - longTextElement.clientHeight;
+    }
+    const rootElement = document.documentElement;
+    rootElement.scrollTop = rootElement.scrollHeight - rootElement.clientHeight;
```
Comment (AlanGreene) [inline packages/components/src/components/Log/Log.js:101]: rather than setting the `scrollTop` directly can we use the `scrollTo` API instead?
i.e. `rootElement.scrollTo({ top: … })`, this also lets us control the scroll behaviour (smooth / auto).

File: packages/components/src/components/Log/Log.js [hunk 4]
```diff
@@ -28,17 +30,147 @@ const itemSize = 15; // This should be kept in sync with the line-height in SCSS
...
+    const logElementRight = (window.innerWidth || document.documentElement.clientWidth) -
+      this.logRef?.current?.getBoundingClientRect().right;
+    document.documentElement.style.setProperty("--log-element-right",
+      `${logElementRight.toString()}px`);
+  };
+
+  getScrollButtons = () => {
+    const { enableLogScrollButtons, intl } = this.props;
+    if (!enableLogScrollButtons) {
+      return <Fragment />;
```
Comment (AlanGreene) [inline packages/components/src/components/Log/Log.js:131]: no need to return a fragment here, `null` is fine

File: packages/components/src/components/Log/Log.js [hunk 5]
```diff
@@ -28,17 +30,147 @@ const itemSize = 15; // This should be kept in sync with the line-height in SCSS
...
+    }
+
+    return (
+      <Fragment>
+        {this.state.isLogTopUnseen ? (
+          <button
```
Comment (AlanGreene) [inline packages/components/src/components/Log/Log.js:137]: Can we use a Carbon `Button` component here? This will ensure consistent styling across future updates.

File: packages/components/src/components/Log/Log.js [hunk 6]
```diff
@@ -28,17 +30,147 @@ const itemSize = 15; // This should be kept in sync with the line-height in SCSS
...
+            key="scroll-to-top-btn"
+            className="scroll-btn bx--copy-btn scroll-to-top-btn"
+            onClick={this.scrollToTopLog}
+            type="button"
+          >
+            <UpToTop16>
+              <title>
+                {intl.formatMessage({
+                  id: 'dashboard.pipelineRun.scrollToTop',
```
Comment (AlanGreene) [inline packages/components/src/components/Log/Log.js:146]: `dashboard.log.scrollToTop` since it's not just for PipelineRuns

File: packages/components/src/components/Log/utils.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
+const scrollRegex = /(auto|scroll)/;
```
Comment (AlanGreene) [inline packages/components/src/components/Log/utils.js:1]: missing copyright / license header

File: packages/components/src/components/Log/utils.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        return false;
+    }
+
+    const horizontalScroll = includeHorizontalScroll ? "" : getStyle(el, "overflow-x");
+    return scrollRegex.test(
+        getStyle(el, "overflow") +
+        getStyle(el, "overflow-y") + horizontalScroll);
+}
+
+const utils = {
```
Comment (AlanGreene) [inline packages/components/src/components/Log/utils.js:51]: we can just export/import the individual functions directly, no need to wrap them in an object here

`export function isElementScrollable(…) {` etc.

File: src/containers/App/App.js [hunk 9]
```diff
@@ -205,6 +205,8 @@ export function App({
   }, [isFetchingConfig, tenantNamespace]);
 
   const logoutButton = <LogoutButton getLogoutURL={() => logoutURL} />;
+  const enableLogAutoScroll = true;
```
Comment (AlanGreene) [inline src/containers/App/App.js:208]: no need for these here, they can just be set in the PipelineRun / TaskRun containers.

File: src/containers/PipelineRun/PipelineRun.js [hunk 10]
```diff
@@ -284,6 +286,8 @@ export /* istanbul ignore next */ function PipelineRunContainer(props) {
         />
       )}
       <PipelineRun
+        enableLogAutoScroll={enableLogAutoScroll}
```
Comment (AlanGreene) [inline src/containers/PipelineRun/PipelineRun.js:289]: no need to accept this as a prop, just set the attribute here
```jsx
<PipelineRun
  enableLogAutoScroll
  enableLogScrollButtons
  error={error}
…
```


same for the TaskRun container

**Review by AlanGreene — CHANGES_REQUESTED** [review]
This doesn't seem to be working for me, I'm not seeing the scroll buttons appear at all in the Dashboard application.

If I edit the Log storybook entries I can see them on the windowed / performance stories but the positioning is not as expected.

Let's split this into 2 PRs as I mentioned before so we can deliver the auto-scroll functionality, then follow up on the scroll buttons separately.

There's also quite a bit of logic involved in determining when to show certain elements / trigger certain behaviour, it would be great to have some unit tests even if it's just for the utils so we have some coverage to give us confidence we're not going to break this in future.

File: packages/components/src/components/Log/Log.js (hunk 5 above)
Comment (MA9521) [reply packages/components/src/components/Log/Log.js:137]: I used the `button` tag because it is already used in https://github.com/tektoncd/dashboard/blob/aec426b0c0d7004dcff1dec627be6ee004b9ea37/packages/components/src/components/LogsToolbar/LogsToolbar.js#L27

File: src/containers/App/App.js (hunk 9 above)
Comment (MA9521) [reply src/containers/App/App.js:208]: my idea was to centralize the flag for the entire app instead of having the potential of a different behaviour in PipelineRun vs TaskRun
